### DISCUSSION TITLE
How to group fields in relation table in view collection?

### DISCUSSION THREAD

**Discussion by iamdineshkumar** [pocketbase/pocketbase#2276] [Q&A]
I have three tables
1. Product with field id, name
2. Invoice with fields id,saleOn(date),items(multiple list of record of invoiceItem table), amount
3. invoiceItem with fields id, product(id of product record), quantity, rate and amount

I need to create a view collection with information of sale_date(based on invoice saleOn), product(product record id from invoiceItem table) and sale_quantity

I came up with below rough syntax but struck up with joining invoiceItem table and it had multiple value and also product should be expanded by sdk for product name
```
SELECT
(ROW_NUMBER() OVER()) as id,
CAST(strftime('%s', strftime('%Y-%m-%dT00:00:00+00:00', saleOn), 'unixepoch') as datetime) as sale_date,
sum(quantity) as sale_quantity
FROM sales
GROUP BY STRFTIME("%d-%m-%Y", saleOn) AND product
```

**Comment by ganigeorgiev** [comment]
Could you provide an example with sample data and expected output?

Additionally, can a single `InvoiceItem` record has more than one `Invoice` records? If No, then you may want to reverse the relation, aka. have a single `relation` field in `InvoiceItem` instead of `multiple` in `Invoice`.

In any case, if you want to apply a condition and join on mutiple `relation` field value, you can use the [`json_each` SQLite function](https://www.sqlite.org/json1.html#jeach), but I'm not sure that I understand your requirements and the described collections structure.

**Reply by iamdineshkumar** [reply]
Product Collection
![image](https://user-images.githubusercontent.com/14805534/231441432-564a2445-a637-48bf-ad93-65c879bdbfee.png)
Invoice
![image](https://user-images.githubusercontent.com/14805534/231442004-0239dfe2-05f4-47ce-98f3-fe89c3edf300.png)

Invoice Items
![image](https://user-images.githubusercontent.com/14805534/231441780-2914080f-fd4f-4762-9b27-a6111d2fb0b8.png)

Invoice have multiple invoice items(so single invoice can have multiple invoice items of different products)

I am trying to achieve is createing view collection which consolidate date wise item sale quantity 

For an ex:
Product Table
| id | name |
| --- | --- |
| 1 | PocketBase Tshirt |
| 2 | Pocker Base Stickers |

itemWiseDailySale (View Collection)
| id | date | product | quantity |
| --- | --- | --- | --- |
| 1 | 2023-04-11 | 1 | 100 |
| 2 | 2023-04-12 | 1 | 50 |
| 3 | 2023-04-12 | 2 | 25 |
 
In above view collection product should be expandable to get product details

**Reply by ganigeorgiev** [reply]
> Invoice have multiple invoice items(so single invoice can have multiple invoice items of different products)

But again, can a single invoice item exist in more than one invoice?
If no, then you may want to reverse the relation, aka. have a single `invoice` `relation` field in `invoiceItems` collection instead of multiple `relation` in `invoice`.

If I'm understanding correctly the `itemWiseDailySale` expected table result, then you can try with a query like:

```sql
SELECT
    (ROW_NUMBER() OVER()) as id,
    invoiceItems.product as product,
    strftime("%Y-%m-%d", saleOn) as saleDate,
    sum(invoiceItems.quantity) as saleQuantity
FROM invoice
INNER JOIN json_each(invoice.items) as je
INNER JOIN invoiceItems ON invoiceItems.id = je.value
GROUP BY saleDate, invoiceItems.product
```

_**edit:** updated the query with the fields from the screenshots_

**Reply by iamdineshkumar** [reply]
Yes there is a possible single invoice item per invoice also... Configured relation of invoiceItems in Invoice is multiple but there is possibility of having one or many invoiceItem per invoice.

**Reply by ganigeorgiev** [reply]
>  Since some invoice can have single item and some may have multiple item

I think there is some confusion. What I meant was whether a single invoice item could exist in multiple different invoices, aka.:

```
invoiceA: [itemA, itemB]
invoiceB: [itemA]
```

If each invoice contains unique invoice items, then you may want to reverse the relation.

**Reply by iamdineshkumar** [reply]
each invoice will have unique Items
invoiceA: [itemA,itemB]
invoiceB:[itemC]

same invoiceItems will not be there in two invoices. So one invoice have one or many invoiceItems but one invoiceItem will be belong to one invoice

**Reply by iamdineshkumar** [reply]
Another important thing which i missed is timezone of date.  Actually to covert to user timezone of sale date need to be considered. 

So, I think it can be done by grouping using strftime
```
SELECT
    (ROW_NUMBER() OVER()) as id,
    invoiceItems.product as product,
    strftime("%Y-%m-%d", saleOn) as saleDate,
    sum(invoiceItems.quantity) as saleQuantity
FROM invoice
INNER JOIN json_each(invoice.items) as je
INNER JOIN invoiceItems ON invoiceItems.id = je.value
GROUP BY strftime("%Y-%m-%d",saleDate,"localtime"), invoiceItems.product
```
